### ISSUE TITLE
[CLI]: Sweep starts many duplicated runs

### ISSUE THREAD

**Issue by andreimargeloiu** [wandb/client#3522]
### Describe the bug

**Description**
I'm a big fan of the sweep functionality. Recently, I noticed that **one [sweep](https://wandb.ai/andreimargeloiu/low-data/sweeps/srhfq97l) spawned many duplicated runs.** In the sweep I perform a grid search across two hyperparameters.

![CleanShot 2022-04-13 at 14 43 19](https://user-images.githubusercontent.com/18227298/163194063-1a38cc80-e655-4c74-972c-1d52b7f40765.png)

I have a random UUID as the experiment tag for all runs inside the sweep, so I'm confident that I gather only runs associated with this sweep.

I use the wandb API to select all runs associated with the UUID and then count the number of runs for each hyper-parameter. For each hyper-parameter, I expected 25 runs (because I run internal cross-validation)

However, the minimum number of runs is 50, and for many of them, the number of duplicated is very high (even 442 runs instead of 25)!

![CleanShot 2022-04-13 at 14 38 59](https://user-images.githubusercontent.com/18227298/163193275-4b4563fb-f17e-4579-8b2c-484f76bc9d28.png)

For example, these four runs ([one](https://wandb.ai/andreimargeloiu/low-data/runs/natkhfjd), [two](https://wandb.ai/andreimargeloiu/low-data/runs/3ufci98x), [three](https://wandb.ai/andreimargeloiu/low-data/runs/36u9wy1x), [four](https://wandb.ai/andreimargeloiu/low-data/runs/1tkxypve)) are identical (the sweep processes launched them multiple times!)

Notice that the config file is identical
![CleanShot 2022-04-13 at 14 47 17](https://user-images.githubusercontent.com/18227298/163195001-819451d5-875d-46c7-aadf-082215805b48.png)

and the losses are identical
![CleanShot 2022-04-13 at 14 48 17](https://user-images.githubusercontent.com/18227298/163195097-e0831e9f-754b-4821-ad93-c2c218e99a06.png)

**Reproduce**
I spawned this [sweep](https://wandb.ai/andreimargeloiu/low-data/sweeps/srhfq97l) from multiple agents, each in a different `tmux` session from the same machine.

What could be the issue?

### Environment

WandB version: 0.12.14
OS: linux
Python 3.7.9

Versions of relevant libraries:
pytorch-lightning==1.5.9

**Comment by armanhar** [comment]
Hey Andrei, thanks for flagging this. I'll try to repro this and file a ticket so that we can address this asap.

**Comment by collinmccarthy** [comment]
I'm having the same issue. I use slurm to launch `wandb agent <SWEEP_ID>` via the CLI interface and I noticed if a job gets pre-empted, when it resumes with the same command it will start a run that is already running.

**Comment by ash0ts** [comment]
Hello @andreimargeloiu and @collinmccarthy ,

Thank you for bringing this to our attention. Our engineers are aware of this issue regarding sweeps that has recently arisen and will be investigating the cause of this.

**Comment by dannygoldstein** [comment]
hi @andreimargeloiu, can you point us to any other sweeps where you saw this behavior? the link above seems to point to a sweep that has since been deleted

**Comment by andreimargeloiu** [comment]
@dannygoldstein sorry, I no longer have access to that sweep.

**Comment by collinmccarthy** [comment]
@dannygoldstein I can give you access to [this sweep](https://wandb.ai/sparse-dnn/sparse-dnn/sweeps/v5kt6bh3) if it helps, but I can't make it public. At the moment my workflow is Slurm -> Elastic launch (DDP) -> wandb.agent on rank 0. I moved away from the CLI due to multiple other issues with this workflow. The only remaining issue is the agent duplicating runs.

When I launch ~25 runs, about half are duplicates (the agent doesn't detect that the sweep is already running, or if it does, it doesn't throw an exception and stop processing like I would expect it to). I'm going to try modifying `wandb.agent` to see if the process-based workflow might work better than threading (not sure if this is possible though).

**Comment by collinmccarthy** [comment]
Here is a `debug-internal.log` where two runs on different machines (on the same shared NFS) have resumed the same run. I assume the logic for chosing whether to launch this run comes from `sender.py:_maybe_setup_resume()`? Are you guys relying on NFS file locks for `wandb.agent` or the CLI? I've had issues with NFS-based locks on our cluster before.

[debug-internal.log](https://github.com/wandb/client/files/8567644/debug-internal.log)
:

**Comment by collinmccarthy** [comment]
Okay, I fixed my issue by switching from the `pyagent` threading based approach to `wandb_agent.run_agent` which calls `AgentProcess` (mostly by commenting out the line where `pyagent` is called). 

At that point the duplicate runs went away but I was still getting some HTTP errors ("500 response executing GraphQL") causing hanging at the beginning. I deleted a bunch of files on Wandb (I was at my storage limit) and then that went away too. I'll just avoid logging the large checkpoint files moving forward to be safe, in case these larger I/O operations are exacerbating timeouts or race conditions. After these changes my sweeps seem to be running correctly (although I haven't tried the `preempted` state again).

I'll try to remember to submit a feature request / PR based on my changes in `wandb_agent.py`. I do really think that DDP -> Multiprocessing Agent on Rank 0 is a great way to support more complicated workflows and still be able to easily debug. And the codebase is already setup to do so.

Thanks

**Comment by ash0ts** [comment]
Hello @collinmccarthy,

I believe your training scripts may be erroring before calling `init()`, leaving Run Queue items in a Leased state, which never gets claimed. This happens when you don't call init() at the beginning of a sweep to ack suggestions. They get put back on the run queue and sent back to the caller after the lease expires. Looking at your sweep controller logs, your suggestions are being re-suggested every ~60 minutes or so, which is how long the lease is.

I am glad to hear that your changes were able to allow your code to run properly. 

Could you provide more details about
> I'll try to remember to submit a feature request / PR based on my changes in wandb_agent.py. I do really think that DDP -> Multiprocessing Agent on Rank 0 is a great way to support more complicated workflows and still be able to easily debug. And the codebase is already set up to do so.

as I can create an internal ticket for our engineers to take a look at.

**Comment by ash0ts** [comment]
Hi @collinmccarthy, since we have not heard back from you we are going to close this request. If you would like to re-open the conversation, please let us know!
​

**Comment by skandermoalla** [comment]
@ash0ts I've just faced this issue recently in a grid sweep with multiple agents in parallel running `wandb` in offline mode.
The runs seems to be duplicated every 60 minutes as mentioned in your previous message, but there are no errors in my runs, they run until completion.
Can this be due to the fact that I'm running offline? During this time the runs show as `crashed` in the wandb project runs, which is very misleading btw.
<img width="1659" alt="Screenshot 2023-06-25 at 21 52 38" src="https://github.com/wandb/wandb/assets/37197319/560ce7b9-2d30-461d-b48b-d00f1fa27209">
<img width="1659" alt="Screenshot 2023-06-25 at 21 52 43" src="https://github.com/wandb/wandb/assets/37197319/c8fdd23b-79d9-4106-8b5e-a5ebcd6bc3e4">

**Comment by michaelmior** [comment]
I seem to be hitting something similar but in my case I have multiple agents waiting on a run with the same run ID, effectively reducing parallelism.

**Comment by kell18** [comment]
I just have basic WanDB setup with ` wandb.init` then `run.config.something = 123` then `wandb.log(metrics)` and it still creates 4 runs, I can see it in the logs, it's so annoying!!!

**Comment by Tokenmw** [comment]
@skandermoalla I also have this issue in offline mode, it will repeat previous K runs results in a loop until the end. Even I update the recently released version. Do you know what's the matter now?

**Comment by alex-schutz** [comment]
This is still proving an issue, a recent offline grid sweep which was supposed to have 320 runs ended up with only 130 (though 320 jobs were successfully launched via slurm). 

Edit: Still experience issues with duplication in offline runs as of 19 Dec 2025

**Comment by koernerfelicia** [comment]
I also have this issue on version `0.23.1`. When I start a sweep with multiple agents, with `run_cap = 30` and a search space of 128 combinations, I only get 21 unique runs.

**Comment by AlejandroCampoy** [comment]
Hi! Is there any update on this issue?

I'm experiencing similar problems on version `0.25.0`. I'm calling the agent from multiple machines in the cloud, and each machine runs several calls in parallel. I've set it up so that the agent is called exactly the same number of times as the total number of possible combinations in the sweep grid.

In the end, the total number of runs executed matches the expected number. However, some hyperparameter configurations are repeated, while other combinations are never executed.

**Update**: My hypothesis is that this duplicated runs bug is preventing the sweep from finishing correctly. As a result, agents can continue requesting runs (more combinations than grid specification), and the sweep keeps returning additional runs with repeated configurations indefinitely.

I'm also using nested parameters in the sweep configuration. Could this potentially be causing any issues?

Would it make sense to open a separate issue for this?
